src/mentorship/ml/models/common.py

```diff
@@ -1,39 +1,94 @@
-    def __init__(self, base_pipeline):
-        self.base_pipeline = base_pipeline
+class RecursiveTSEstimator(BaseEstimator):
+    def __init__(self, split_key, target_col, base_pipelines=None, zero_categories=None):
+        if zero_categories is None:
+            zero_categories = []
+        if base_pipelines is None:
+            base_pipelines = {}
+        self.split_key = split_key
+        self.target_col = target_col
+        self.base_pipelines = base_pipelines
+        self.zero_categories = zero_categories
         self.pipelines_ = {}
 
-    def fit(self, X, y):
-        for family in X['family'].unique():
-            indexer = X[X['family'] == family].index
-            X_part = X[X['family'] == family]
-            y_part = y.loc[indexer]
+    def fit(self, X, y, **fit_params):
+        for key, X_part in X.groupby(self.split_key):
+            y_part = y.loc[X_part.index]
 
-            pipeline = clone(self.base_pipeline)
-            pipeline = pipeline.fit(X_part, y_part)
+            # save train data for using it in counting rolling features for the test data in the future
+            self.base_pipelines[key].train_data = X_part
 
-            self.pipelines_[family] = pipeline
+            pipeline = clone(self.base_pipelines[key])
+
+            if pipeline.use_final_metric:
+                X_part[pipeline.target_col] = np.log(X_part[pipeline.target_col] + 1)
+                y_part = np.log(y_part + 1)
+
+            pipeline = pipeline.fit(X_part.drop(columns=[self.split_key]), y_part)
+
+            self.pipelines_[key] = pipeline
         return self
 
     def predict(self, X):
-        X.loc[:, 'forecast'] = 0
-        for family in X['family'].unique():
-            X_part = X[X['family'] == family]
-            X_part = X_part.drop('forecast', axis=1)
-            pipeline = self.pipelines_[family]
-            X.loc[X_part.index, 'forecast'] = pipeline.predict(X_part)
-
-        y_pred = X['forecast']
-        X = X.drop(columns='forecast')
-        return y_pred
+        y_preds = {}
+        for split_key_value, X_part in X.groupby(self.split_key):
+            if split_key_value in self.zero_categories:
+                y_preds[split_key_value] = pd.Series(0, index=X_part.index, name='forecast')
+                continue
 
-    def get_params(self, deep=True):
-        return {'base_pipeline': self.base_pipeline}
+            pipeline = self.pipelines_[split_key_value]
 
-    def set_params(self, **parameters):
-        for parameter, value in parameters.items():
-            setattr(self, parameter, value)
-        return self
+            # all train set + processed test days (for counting rolling features)
+            X_part_all_rows = pipeline.train_data
+
+            # sort dates of X_part for safety
+            sorted_dates = X_part[pipeline.date_column].unique().tolist()
+            sorted_dates.sort()
+
+            for current_day_number, current_day in enumerate(sorted_dates):
```

Comment:
Some logic inside this loop repeats the logic of generating lags and rolling features. I would suggest you rewriting it in the following way:
1. Add some method like `update_recent_history` to `LagComputer` that will concatenate the existing part of the train set with the new portion of data. And possibly truncates the old and outdated days in the part of the existing train.
2. At the beginning of each iteration apply the full pipeline to the data corresponding to the current day. It will prepare all required features for the model.
3. At the end of each iteration call `update_recent_history` method and pass the predicted dataframe to it.